File: sorting/QuickSort.java
package sorting;

import java.util.*;

public class QuickSort {

    public static void main(String[] args) {
        int[] a1 = {1,2,13,9,5,6,10,5};
        sort(a1,0,a1.length-1);
        for (int i : a1) {
            System.out.print(i + " ");
        }
        System.out.println();
        int[] a2 = {1,2,-13,9,50,6,10,5};
        sort(a2,0,a2.length-1);
        for (int i : a2) {
            System.out.print(i + " ");
        }
    }

    public static void sort(int[] arr, int low, int high) {
        if (high - low < 1) {
            //skip
        } else {
            int split = partition(arr, low, high);
            sort(arr,low,split-1);
            sort(arr,split,high);
        }
    };

    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[low + ((high - low) / 2)];
        int i = low;
        int j = high;
        while (i <= j) {
            if (arr[i] >= pivot && arr[j] <= pivot) {
                int tmp = arr[i];
                arr[i] = arr[j];
                arr[j] = tmp;
                i++;
                j--;
            } else if (arr[i] < pivot) {
                i++;
            } else {
                j--;
            }
        }
        return i;
    }

    public static void print(int[] arr) {
        System.out.println();
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }
}

File: sorting/python/quicksort.py
def sort(arr, start, end):
    if start >= end:
        return
    pivot = (start+end)/2
    i = start
    j = end
    while i < j:
        if arr[i] >= arr[pivot] and arr[j] <= arr[pivot]:
            tmp = arr[i]
            arr[i] = arr[j]
            arr[j] = tmp
            i+=1
            j-=1
        else:
            if arr[i] < arr[pivot]:
                i+=1
            if arr[j] > arr[pivot]:
                j-=1
    sort(arr,start,pivot)
    sort(arr,pivot+1,end)




a1 = [12,5,6,2,4,13]
sort(a1,0,len(a1)-1)
print a1

a2 = [12,5,4,13]
sort(a2,0,len(a2)-1)
print a2

a3 = [12,5]
sort(a3,0,len(a3)-1)
print a3

a4 = [5]
sort(a4,0,len(a4)-1)
print a4

a5 = [12,5,5,12]
sort(a5,0,len(a5)-1)
print a5


